sample/src/androidTest/kotlin/com/slack/circuit/sample/petdetail/PetDetailTest.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2022 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.circuit.sample.petdetail
+
+import androidx.activity.ComponentActivity
+import androidx.compose.ui.test.assertIsDisplayed
+import androidx.compose.ui.test.assertTextEquals
+import androidx.compose.ui.test.junit4.createAndroidComposeRule
+import androidx.compose.ui.test.onNodeWithTag
+import androidx.compose.ui.test.onNodeWithText
+import com.google.common.truth.Truth.assertThat
+import com.slack.circuit.Circuit
+import com.slack.circuit.CircuitProvider
+import com.slack.circuit.sample.R
+import com.slack.circuit.sample.petdetail.PetDetailTestConstants.ANIMAL_CONTAINER_TAG
+import com.slack.circuit.sample.petdetail.PetDetailTestConstants.PROGRESS_TAG
+import com.slack.circuit.sample.petdetail.PetDetailTestConstants.UNKNOWN_ANIMAL_TAG
+import com.slack.circuit.sample.petdetail.PetPhotoCarouselTestConstants.CAROUSEL_TAG
+import org.junit.Rule
+import org.junit.Test
+
+class PetDetailTest {
+  @get:Rule val composeTestRule = createAndroidComposeRule<ComponentActivity>()
+
+  @Test
+  fun petDetail_show_progress_indicator_for_loading_state() {
+    composeTestRule.run {
+      setContent { PetDetail(PetDetailScreen.State.Loading) }
+
+      onNodeWithTag(PROGRESS_TAG).assertIsDisplayed()
+      onNodeWithTag(UNKNOWN_ANIMAL_TAG).assertDoesNotExist()
+      onNodeWithTag(ANIMAL_CONTAINER_TAG).assertDoesNotExist()
+    }
+  }
+
+  @Test
+  fun petDetail_show_message_for_unknown_animal_state() {
+    composeTestRule.run {
+      setContent { PetDetail(PetDetailScreen.State.UnknownAnimal) }
+
+      onNodeWithTag(PROGRESS_TAG).assertDoesNotExist()
+      onNodeWithTag(ANIMAL_CONTAINER_TAG).assertDoesNotExist()
+
+      onNodeWithTag(UNKNOWN_ANIMAL_TAG)
+        .assertIsDisplayed()
+        .assertTextEquals(activity.getString(R.string.unknown_animals))
+    }
+  }
+
+  @Test
+  fun petDetail_show_animal_for_success_state() {
+    val success =
+      PetDetailScreen.State.Success(
+        url = "url",
+        photoUrls = listOf(""), // using empty string here to trigger use of fallback image
+        photoUrlMemoryCacheKey = null,
+        name = "Baxter",
+        description = "Grumpy looking Australian Terrier"
+      )
+
+    var carouselScreen: PetPhotoCarouselScreen? = null
+    val circuit =
```

Comment:
While this works, it does require a bit of plumbing to get working. If this is a pattern we expect will be commonly used, we might want to hide this in some test harness helper logic

Reply:
good callout